Add lowering for BF16 arith/math operations to F32 with truncf/extf wrappers

rsuderman opened the pull request:
We can support BF16 execution by casting to F32 types and performing the F32 math in that space. This allows bf16 support once we have a simplified conversion betweeen bf16 and f32. For now this path is included after the BF16 to F32 promotion pass so it will not trigger. A follow up will fully enable BF16 support and disable the F32 conversion.

rsuderman committed:
Added support for BF16 execution on CPU

We can support BF16 execution by casting to F32 types and performing
the F32 math in that space. This gives the benefit of using BF16 as
the executable type while still supporting on non BF16 hardware.

jpienaar reviewed (commented):
Nice!

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp:
```diff
@@ -34,6 +36,42 @@ namespace IREE {
+  Type inputETy = getElementTypeOrSelf(inputs[0].getType());
+  if (!getElementTypeOrSelf(type).isa<FloatType>()) return nullptr;
+
+  if (inputETy.getIntOrFloatBitWidth() > eTy.getIntOrFloatBitWidth())
```
Nit: Please add braces. (this diverges from LLVM style but is recommended in IREE)

rsuderman replied:
Fixed.

MaheshRavishankar commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:
```diff
@@ -685,6 +686,7 @@ static void addLowerToLLVMPasses(OpPassManager &passManager) {
     passManager.addNestedPass<func::FuncOp>(
         createLLVMCPUEmitVectorizationRemarksPass());
   }
+  passManager.addPass(IREE::Util::createPromoteArithBF16ToF32Pass());
```
I am wondering if this is enough. Arent the operations in the body `linalg` operations with `bf16` types? I am wondering why this is different compared to the `i1` -> `i8` handling?

jpienaar replied:
Do we/could we add a test that verifies this?

rsuderman replied:
So unlikely the `i1` to `i8` conversion we do not change the underlying storage buffers, rather they remain as a 2-byte per element buffer. What does change is we perform casts surrounding all of the `arith` operations so that they hit known to work versions of the ops (e.g. `f32`). 

This occurs after the conversion to `scf` but should occur both within the dispatched for loops and the VM code, as we apply the conversion to any numerical operations throughout the model.

I am still waiting on the corresponding LLVM change to integrate as it missed the most recent integration cycle. Once it is in we should be able to introduce an execution example. For now we still perform the raw `bf32` to `i32` conversion on `mhlo` models.

rsuderman replied:
We may be able to do some similar tricks for `i1` packing as well. I imagine `exti` could have a similar conversion where we extract only the bits needed, though the loading semantics would need to be adjusted.

MaheshRavishankar replied:
ok I see... well it makes me a bit concerned that we have two places where we are adjusting types... but it is true, one is at buffer level and another is at register level....

rsuderman committed:
Cleanup brackes lit issues

rsuderman committed:
Disable existing BF16-to-F32 conversion and add end-to-end test

We can support BF16 directly without converting F32 except at arith
operations.

rsuderman committed:
Fix bazel build file for excludes

MaheshRavishankar reviewed (approved):
Looks good! 

@powderluv I think this will enable running stable diffusion on CPU backends now.

rsuderman committed:
buildifier

jpienaar commented on tests/e2e/xla_ops/BUILD.bazel:
```diff
@@ -122,6 +124,7 @@ iree_check_single_backend_test_suite(
             "cosine.mlir",
             "divide.mlir",
             "dot.mlir",
+            # "dot_bf16.mlir",
```
Should we just remove here?

rsuderman replied:
Removed.

rsuderman committed:
Removed commented out files

rsuderman committed:
Fix illegal cast issue for arith conversion.

rsuderman committed:
Fix risc/arm builds missing bf16-f32 interop
